llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/IREEPyDM/IR/IREEPyDMOps.td

```diff
@@ -0,0 +1,521 @@
+
+def IREEPyDM_SelectOp : IREEPyDM_PureOp<"select", [
+  AllTypesMatch<["true_value", "false_value", "result"]>
+    ]> {
+  let summary = "Select a true or false value based on condition";
+  let arguments = (ins
+      IREEPyDM_BoolType:$condition,
+      IREEPyDM_AnyValueType:$true_value,
+      IREEPyDM_AnyValueType:$false_value);
+  let results = (outs IREEPyDM_AnyValueType:$result);
+  let assemblyFormat = [{
+    $condition `,` $true_value `,` $false_value `:` type($result) attr-dict
+  }];
+}
+
+def IREEPyDM_ExprStatementDiscardOp : IREEPyDM_Op<"expr_statement_discard"> {
+  let summary = "Anchors an expression evaluated as a statement";
+  let description = [{
+    This op is used early during import to keep a statement-evaluated expression
+    live until more effect information is available to anchor it properly.
+  }];
+  let arguments = (ins IREEPyDM_AnyValueType:$value);
+  let assemblyFormat = [{
+    $value `:` type($value) attr-dict
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Computation
+//===----------------------------------------------------------------------===//
+
+def IREEPyDM_DynamicBinaryPromoteOp : IREEPyDM_PureOp<"dynamic_binary_promote"> {
+  let summary = "Promotes two arguments that may be of numeric types";
+  let description = [{
+    Takes two values of arbitrary type which may be input to a following
+    binary arithmetic operation. If they are both numeric, returns the two
+    arguments, promoted to the bounding, common numeric type. Otherwise,
+    returns the inputs as-is.
+
+    This op will always produce boxed, type erased values since that is always
+    legal for runtime evaluation. However, for known input types, it will
+    canonicalize to various identities or fixed numerics, making the trailing
+    boxing trivial to type propagate past.
+
+    Note that this operation could be implemented with language-level pattern
+    matching, but it is both very frequently used and complicated from a
+    type system perspective. As such, we make it a black box with custom
+    canonicalization and lowering.
+  }];
+
+  let arguments = (ins
+      IREEPyDM_AnyValueType:$left,
+      IREEPyDM_AnyValueType:$right);
+  let results = (outs
+      IREEPyDM_GenericObjectType:$left_prime,
+      IREEPyDM_GenericObjectType:$right_prime);
+  let assemblyFormat = [{
+    $left `,` $right `:` type($left) `,` type($right) attr-dict
+  }];
+}
+
+def IREEPyDM_ApplyBinaryOp : IREEPyDM_PureOp<"apply_binary"> {
+  let summary = "Applies one of Python's binary operations";
+  let description = [{
+    The operation to apply is specified as per the data model:
+      https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types
+    Here it is presented with leading and trailing double underscores (i.e.
+    "add", "sub", etc).
+
+    Numeric types must be promoted to a common type prior to application.
+  }];
+  let arguments = (ins
+    StrAttr:$dunder_name,
```

Comment:
had not heard "dunder" before - made me think of 
![image](https://user-images.githubusercontent.com/75337/131374479-c98f8a7a-acdf-4630-89e0-54f4d9f35c4a.png)

Reply:
Haha - is it bad that I've been using the Pythonic "dunder" for so long, this *didn't* pop into my head?